llvm/lib/CodeGen/FPBuiltinFnSelection.cpp

```diff
@@ -77,12 +109,48 @@ static bool selectFnForFPBuiltinCalls(const TargetLibraryInfo &TLI,
   StringSet<> RecognizedAttrs = {FPBuiltinIntrinsic::FPBUILTIN_MAX_ERROR};
   if (BuiltinCall.hasUnrecognizedFPAttrs(RecognizedAttrs)) {
     report_fatal_error(
-        Twine(BuiltinCall.getCalledFunction()->getName()) +
+        Twine(OldFuncName) +
             Twine(" was called with unrecognized floating-point attributes.\n"),
         false);
     return false;
   }
 
+  Triple T(BuiltinCall.getModule()->getTargetTriple());
+  llvm::Triple::ArchType Arch = T.getArch();
+  // Several functions for "sycl" and "cuda" requires "0.5" accuracy levels,
+  // which means correctly rounded results. For now x86 host AltMathLibrary
+  // doesn't have such ability. For such accuracy level, the fpbuiltins
+  // should be replaced by equivalent IR operation or llvmbuiltins.
+  if ((Arch == Triple::x86 || Arch == Triple::x86_64) &&
+      BuiltinCall.getRequiredAccuracy().value() == 0.5) {
+    if (OldFuncName.contains("fadd")) {
+      return replaceWithLLVMIR(BuiltinCall, "fadd");
+    } else if (OldFuncName.contains("fsub")) {
+      return replaceWithLLVMIR(BuiltinCall, "fsub");
+    } else if (OldFuncName.contains("fmul")) {
+      return replaceWithLLVMIR(BuiltinCall, "fmul");
+    } else if (OldFuncName.contains("fdiv")) {
+      return replaceWithLLVMIR(BuiltinCall, "fdiv");
+    } else if (OldFuncName.contains("frem")) {
+      return replaceWithLLVMIR(BuiltinCall, "frem");
+    } else if (OldFuncName.contains("sqrt")) {
+      OldFuncName.consume_front("llvm.fpbuiltin");
+      std::string LLVMIntrinsicName = "llvm" + OldFuncName.str();
+      return replaceWithAltMathFunction(BuiltinCall, LLVMIntrinsicName);
```

Comment:
This isn't the right way to do this. You should be using IRBuilder::CreateIntrinsic() with Intrinsic::sqrt (here) or Intrinsic::ldexp (below). You will probably want a new helper function to do that. What you've done here will work now, but it could break in the future if the intrinsic handling or the naming for either of these changes.